File: TRPO/lab6/myproject/src/main/kotlin/com/example/app/MainApp.kt
package com.example.app

import tornadofx.*
import javafx.collections.FXCollections
import javafx.beans.value.ObservableValue
import java.io.File
import com.example.factory.UserFactory
import com.example.model.UserType
import com.example.utils.SinglyLinkedList

class MainApp : App(MainView::class)

class MainView : View("Singly List Manager") {
    private val factory = UserFactory()
    private var list = SinglyLinkedList(factory.getBuilderByName("GPSCoordinate"))
    private val items = FXCollections.observableArrayList<String>()

    // Элементы управления GUI
    private val typeCombo = combobox<String> {
        items.setAll(factory.getTypeNames())
        selectionModel.selectFirst()
        // Исправление: явное указание типов в слушателе
        selectionModel.selectedItemProperty().addListener { 
            _: ObservableValue<out String>, 
            _: String?, 
            _: String? -> 
            updateListType() 
        }
    }


    private val inputField = textfield { 
        promptText = "Введите данные (например, 55.751244;37.618423;2023-10-01T12:00:00)" 
    }

    private val indexField = textfield { promptText = "Индекс" }
    private val listView = listview(items)

    override val root = vbox {
        spacing = 10.0
        
        hbox {
            spacing = 10.0
            add(typeCombo)
            add(inputField)
            add(indexField)
            
            vbox {
                spacing = 5.0
                button("Добавить").action { addItem() }
                button("Получить").action { getItem() }
                button("Удалить").action { removeItem() }
                button("Сортировать").action { sortList() }
                button("Сохранить").action { saveToFile() }
                button("Загрузить").action { loadFromFile() }
            }
        }
        
        add(listView)
    }

    // Обновление типа данных списка
    private fun updateListType() {
        list = SinglyLinkedList(factory.getBuilderByName(typeCombo.value))
        updateListView()
    }

    // Обновление отображаемого списка
    private fun updateListView() {
        items.setAll(list.toList().mapIndexed { idx, item -> "$idx: $item" })
    }

    // Обработчики событий
    private fun addItem() {
        try {
            val value = list.userType.parseValue(inputField.text)
            list.add(value)
            inputField.clear()
            updateListView()
        } catch (e: Exception) {
            error("Ошибка", "Неверный формат: ${e.message}")
        }
    }

    private fun getItem() {
        try {
            val index = indexField.text.toInt()
            val value = list.get(index)
            information("Значение", value?.toString() ?: "Не найдено")
            indexField.clear()
        } catch (e: Exception) {
            error("Ошибка", "Неверный индекс: ${e.message}")
        }
    }

    private fun removeItem() {
        try {
            val index = indexField.text.toInt()
            if (list.remove(index)) {
                updateListView()
                indexField.clear()
            } else {
                error("Ошибка", "Индекс $index не существует")
            }
        } catch (e: Exception) {
            error("Ошибка", "Некорректный индекс: ${e.message}")
        }
    }

    private fun sortList() {
        list.sort()
        updateListView()
        information("Успех", "Список отсортирован")
    }

    private fun saveToFile() {
        try {
            list.saveToFile("data.bin")
            information("Успех", "Данные сохранены в data.bin")
        } catch (e: Exception) {
            error("Ошибка", "Ошибка сохранения: ${e.message}")
        }
    }

    private fun loadFromFile() {
        try {
            list = SinglyLinkedList(list.userType).apply { loadFromFile("data.bin") }
            updateListView()
            information("Успех", "Данные загружены из data.bin")
        } catch (e: Exception) {
            error("Ошибка", "Ошибка загрузки: ${e.message}")
        }
    }
}

File: TRPO/lab5/test/lab5/src/main/scala/MainApp.scala
import scalafx.application.JFXApp
import scalafx.scene.Scene
import scalafx.scene.control._
import scalafx.scene.layout.{VBox, HBox}
import scalafx.collections.ObservableBuffer
import scalafx.scene.control.Alert.AlertType
import java.io.File

object MainApp extends JFXApp {
  private var currentList = new SinglyLinkedList(new IntegerType)
  private val factory = new UserFactory

  // Объявляем элементы управления как поля класса
  private var listView: ListView[String] = _
  private var inputField: TextField = _
  private var indexField: TextField = _

  stage = new JFXApp.PrimaryStage {
    title = "Singly List Manager"
    scene = new Scene(800, 600) {
      // Инициализация элементов
      val typeCombo = new ComboBox[String](factory.getTypeNameList)
      inputField = new TextField { promptText = "Значение" }
      indexField = new TextField { promptText = "Индекс" }
      val addButton = new Button("Добавить")
      val getButton = new Button("Получить")
      val removeButton = new Button("Удалить")
      val sortButton = new Button("Сортировать")
      val saveButton = new Button("Сохранить")
      val loadButton = new Button("Загрузить")
      listView = new ListView[String]()

      // Разметка
      root = new VBox {
        children = Seq(
          new HBox {
            children = Seq(
              typeCombo,
              inputField,
              indexField,
              new VBox {
                children = Seq(
                  addButton,
                  getButton,
                  removeButton,
                  sortButton,
                  saveButton,
                  loadButton
                )
              }
            )
          },
          listView
        )
      }

      // Обработчики событий
      addButton.onAction = _ => handleAdd()
      getButton.onAction = _ => handleGet()
      removeButton.onAction = _ => handleRemove()
      sortButton.onAction = _ => handleSort()
      saveButton.onAction = _ => handleSave()
      loadButton.onAction = _ => handleLoad()
      
      typeCombo.onAction = _ => {
        currentList = new SinglyLinkedList(factory.getBuilderByName(typeCombo.value()))
        updateListView()
      }
    }
  }

  private def updateListView(): Unit = {
    listView.items = ObservableBuffer.from(
      currentList.toList.zipWithIndex.map { case (item, idx) => 
        s"$idx: ${item.toString}"
      }
    )
  }

  private def handleAdd(): Unit = try {
    val value = currentList.userType.parseValue(inputField.text())
    currentList.add(value)
    updateListView()
  } catch {
    case e: Exception => showError("Ошибка добавления", e)
  }

  private def handleGet(): Unit = try {
    val index = indexField.text().toInt
    currentList.get(index) match {
      case Some(value) => showInfo(s"Значение по индексу $index: ${value.toString}")
      case None => showError("Неверный индекс", new Exception("Индекс вне диапазона"))
    }
  } catch {
    case e: Exception => showError("Ошибка получения", e)
  }

  private def handleRemove(): Unit = try {
    val index = indexField.text().toInt
    if (currentList.remove(index)) updateListView()
    else showError("Неверный индекс", new Exception("Индекс вне диапазона"))
  } catch {
    case e: Exception => showError("Ошибка удаления", e)
  }

  private def handleSort(): Unit = {
    currentList.sort()
    updateListView()
  }

  private def handleSave(): Unit = {
    val file = new File("list.dat")
    currentList.saveToFile(file.getAbsolutePath)
    showInfo("Список сохранен в list.dat")
  }

  private def handleLoad(): Unit = try {
    val file = new File("list.dat")
    currentList = currentList.loadFromFile(file.getAbsolutePath)
    updateListView()
    showInfo("Список загружен из list.dat")
  } catch {
    case e: Exception => showError("Ошибка загрузки", e)
  }

  private def showError(message: String, e: Exception): Unit = {
    new Alert(AlertType.Error) {
      title = "Ошибка"
      headerText = s"$message: ${e.getMessage}"
    }.showAndWait()
  }

  private def showInfo(message: String): Unit = {
    new Alert(AlertType.Information) {
      title = "Информация"
      headerText = message
    }.showAndWait()
  }
}